### PR TITLE
Add basic user creation and auth

### PR DESCRIPTION
**Pull request by trueb2**
This PR also adds CI to show that nothing is broken.

Goals:
I want the endpoint to be protected by a bearer token that is managed internally. In the future, I may want to expire individual or batch tokens.

Tasks:

- [x] Create user with username and password
- [x] Store bcrypted hash string of password as internal token
- [x] Expose symmetrically encrypted hash string as token for user that expires on secret or password change
- [x] Lookup user from a symmetrically decrypted token and verify bcrypted hash password matches
- [x] Access protected routes after verifying token
- [x] Allow user to change own password
- [x] Out of Scope: Roles and Policies

Specific TODOs:

- [x] Limit user to update only self during req/token verification
- [x] Configure seed data for the database and remove alternative validator paths
- [ ] Wrap crypto errors in enum and produce proper error status and message
- [ ] Replace String usages with &str, array, and slices where possible in token manipulation

Example usage:

```
jwtrueb@dhcp-10-101-176-87 asset-api % echo '{ "username": "jacob", "password": "qsib2" }' | http post :6001/users 'Authorization: Bearer A841BE66-84AC-4BA7-B0E1-D34B1FC2F08A'
HTTP/1.1 200 OK
content-length: 84
content-type: application/json
date: Fri, 06 Nov 2020 01:43:55 GMT

{
    "id": 45,
    "token": "w3uWMpF8u5BtHPzL+JxriRq6PlZ895cUkuKNGC0504vAJR5M+qj9kFEgVnS/JO/N"
}
jwtrueb@dhcp-10-101-176-87 asset-api % echo '{ "username": "jacob", "password": "foo2" }' | http put :6001/users/40 'Authorization: Bearer LhxDXrCsqYD9Ja0DUocbcF+OhGc/dVwWocX8sx7vVrYffNRn4F+VpwsUOx6KvYYt'
HTTP/1.1 200 OK
content-length: 84
content-type: application/json
date: Fri, 06 Nov 2020 01:44:20 GMT

{
    "id": 40,
    "token": "76eVDRNgIurDVxPJ1DTSjIoQD2ZChIv+oxie+aj6U8Sqarvqp7Vopk2YkQ/9KJJF"
}
jwtrueb@dhcp-10-101-176-87 asset-api % http :6001/asset_tags 'Authorization: Bearer LhxDXrCsqYD9Ja0DUocbcF+OhGc/dVwWocX8sx7vVrYffNRn4F+VpwsUOx6KvYYt'      
HTTP/1.1 401 Unauthorized
content-length: 0
date: Fri, 06 Nov 2020 01:45:03 GMT
www-authenticate: Bearer


jwtrueb@dhcp-10-101-176-87 asset-api % http :6001/asset_tags 'Authorization: Bearer 76eVDRNgIurDVxPJ1DTSjIoQD2ZChIv+oxie+aj6U8Sqarvqp7Vopk2YkQ/9KJJF'                                                       
HTTP/1.1 200 OK
content-length: 2
content-type: application/json
date: Fri, 06 Nov 2020 01:45:14 GMT

[]
```

### CODE REVIEW

File: src/auth.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            } else {
+                let config = req.app_data::<Config>()
+                    .map(|data| data.clone())
+                    .unwrap_or_else(Default::default);
+
+                Err(AuthenticationError::from(config).into())
+            }
+        },
+        token => {
+            match users::User::find_by_token(String::from(token)) {
```
Comment (trueb2) [inline src/auth.rs:24]: Hitting the User route requires 3 separate database lookups with the way things are now. Additionally, we create a bcrypt hash string when creating a new user. Latency looked to be about 50ms to create a new user and 2ms to access a protected endpoint.

File: src/db.rs [hunk 2]
```diff
@@ -14,7 +14,7 @@ embed_migrations!();
 lazy_static! {
     static ref POOL: Pool = {
         let (db_url, pool_size) = match cfg!(test) {
-            true => (String::from("postgres://localhost/asset_api"), 1),
+            true => (String::from("postgres://postgres:postgres@localhost/asset_api"), 1),
```
Comment (trueb2) [inline src/db.rs:17]: CI is setup for postgres:postgres. You can set this up in your local postgres database with the `\password` command. Alternatively, you can just use docker and not install anything.
..

File: src/error_handler.rs [hunk 3]
```diff
@@ -38,6 +39,47 @@ impl From<DieselError> for CustomError {
...
+impl From<actix_web::error::Error> for CustomError {
+    fn from(_error: actix_web::error::Error) -> CustomError {
+        CustomError { error_message: String::from("Bad request"), error_status_code: 400 }
+    }
+}
+impl From<actix_web_httpauth::extractors::AuthenticationError<actix_web_httpauth::headers::www_authenticate::bearer::Bearer>> for CustomError {
+    fn from(_error: actix_web_httpauth::extractors::AuthenticationError<actix_web_httpauth::headers::www_authenticate::bearer::Bearer>) -> CustomError {
+        CustomError { error_message: String::from("Bad request"), error_status_code: 400 }
+    }
+}
```
Comment (trueb2) [inline src/error_handler.rs:81]: These are not necessarily 400s or 501s. If we get a bad token, then we should through some error and give a 401 back to the user. Therefore, these codes are not correct, but they do prevent unauthorized access

File: src/users/routes.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        let user = match User::find_by_token(String::from(bearer_auth.token())) {
+            Ok(user) => user,
+            Err(error) => {
+                match error.error_status_code {
+                    404 => {
+                        return err(CustomError::new(401, String::from("Unauthorized")))
+                    },
+                    _ => {
+                        return err(Self::Error::from(error))
+                    }
```
Comment (trueb2) [inline src/users/routes.rs:31]: Here is an example of producing the correct error code for the encountered error considering the situation.

File: migrations/2020-10-26-162809_create_users/up.sql [hunk 5]
```diff
@@ -4,7 +4,7 @@ CREATE TABLE users
 (
     id BIGSERIAL PRIMARY KEY,
     username TEXT NOT NULL,
-    hashed_password TEXT NOT NULL,
+    token TEXT NOT NULL,
```
Comment (trueb2) [inline migrations/2020-10-26-162809_create_users/up.sql:7]: Once we are deployed, this type of change will be unacceptable. We will have to add migrations to alter the table. In the meantime, we can drop and create a new db.

**Review by xunchuan-liu — APPROVED** [review]
Merging into master before beginning my workflow

**Pull request merged**